cmd/cloud_sql_proxy/internal/healthcheck/healthcheck.go

```diff
@@ -132,20 +135,56 @@ func isLive() bool {
 	return true
 }
 
-// isReady will check the following criteria before determining whether the
-// proxy is ready for new connections.
+// isReady will check the following criteria:
 // 1. Finished starting up / been sent the 'Ready for Connections' log.
-// 2. Not yet hit the MaxConnections limit, if applicable.
+// 2. Not yet hit the MaxConnections limit, if set.
+// 3. Able to dial all specified instances without error.
 func isReady(c *proxy.Client, s *Server) bool {
-	// Not ready until we reach the 'Ready for Connections' log
+	// Not ready until we reach the 'Ready for Connections' log.
 	if !s.proxyStarted() {
-		logging.Errorf("Readiness failed because proxy has not finished starting up.")
+		logging.Errorf("[Health Check] Readiness failed because proxy has not finished starting up.")
 		return false
 	}
 
 	// Not ready if the proxy is at the optional MaxConnections limit.
 	if !c.AvailableConn() {
-		logging.Errorf("Readiness failed because proxy has reached the maximum connections limit (%d).", c.MaxConnections)
+		logging.Errorf("[Health Check] Readiness failed because proxy has reached the maximum connections limit (%v).", c.MaxConnections)
+		return false
+	}
+
+	// Not ready if one or more instances cannot be dialed.
+	instances := s.instances
+	if s.instances == nil { // Proxy is in fuse mode.
+		instances = c.GetInstances()
+	}
+
+	ctx := context.Background()
+
+	canDial := true
+	dialL := sync.Mutex{}
+	var wg sync.WaitGroup
+	wg.Add(len(instances))
```

Comment:
Let's move this into the body of the for loop and make it: `wg.Add(1)` for each iteration of the loop a-la https://pkg.go.dev/sync#example-WaitGroup.